eessi/testsuite/tests/apps/walberla/walberla.py

```diff
@@ -0,0 +1,153 @@
+from eessi.testsuite.eessi_mixin import EESSI_Mixin
+from eessi.testsuite.utils import find_modules
+
+
+def filter_scales():
+    """
+    Filtering function for filtering scales for the waLBerla test returns all scales from 8 cores to 2 nodes. This is
+    because this case is testing strong scaling and the number of cells per MPI process become too less typically once
+    it crosses 2 nodes. Further filtering is done within the test based on MPI tasks. The code is capable of running on
+    1 to 4 cores as well but it will take about 1 hour to run on a single core.
+    """
+    return [
+        k for (k, v) in SCALES.items()
+        if ((v['num_nodes'] <= 2) and (v.get('node_part', 0) != 0)) or (v.get('num_cpus_per_node', 0) > 4)
+    ]
+
+
+@rfm.simple_test
+class EESSI_WALBERLA_BACKWARD_FACING_STEP(rfm.RunOnlyRegressionTest, EESSI_Mixin):
+    """
+    This is the main walberla class for the backward facing step test. The test consists of steps that modify the input
+    file from the precompiled walberla tutorials binary. These are run as pre-run commands and the main test with the
+    executable `05_BackwardFacingStep` is measured for performance.
+    The code will run on one core but will take about an hour to run therefore this test is implemented to execute on
+    at least 8 cores and at most 384 cores.
+    """
+    executable = '05_BackwardFacingStep'
+    executable_opts = ['05_BackwardFacingStep.prm']
+    time_limit = '30m'
+    readonly_files = ['']
+    device_type = parameter([DEVICE_TYPES.CPU])
+    module_name = parameter(find_modules('waLBerla'))
+    valid_systems = ['*']
+    scale = parameter(filter_scales())
+
+    @run_after('init')
+    def set_compute_unit(self):
+        """
+        This test is a CPU only test.
+        Set the compute unit to which tasks will be assigned:
+        one task per CPU core for CPU runs.
+        """
+        if self.device_type == DEVICE_TYPES.CPU:
+            self.compute_unit = COMPUTE_UNITS.CPU
+        else:
+            msg = f"No mapping of device type {self.device_type} to a COMPUTE_UNITS was specified in this test"
+            raise NotImplementedError(msg)
+
+    def required_mem_per_node(self):
+        return (400 + self.num_tasks_per_node * 300)
+
+    @run_after('setup')
+    def check_core_count(self):
+        # The walberla backward facing step test case can run on a max of 384 cores and a minimum of 8 cores to run
+        # within reasonable time.
+        if self.num_tasks > 384:
+            self.skip(msg="The maximum number of cores that the strong scaling can hold for this test is 384. Launch on"
+                      "a scale with a smaller core count.")
+        elif self.num_tasks < 8:
+            self.skip(msg="The minimum number of MPI tasks required for this test is 8 to run within reasonable time."
+                      "Launch of scale with larger core count")
+
+    @run_before('run')
+    def prepare_environment(self):
```

Comment:
This should probably get a docstring. I'd keep it more high-level than the detailed sed-explaination, but just mention what it does ("modification of input files so that this example runs on multinode", I guesss?)